### DISCUSSION TITLE
Text editor improvements

### DISCUSSION THREAD

**Discussion by Rytelier** [godotengine/godot-proposals#7386] [GUI]
### Describe the project you are working on

Code, shaders, database

### Describe the problem or limitation you are having in your project

Some text editing task are becoming tedious, like quick cloning lines, renaming objects, selecting multiple items.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Text editor got good improvements over time, there's more functionality though that would help writing text based resources efficiently.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

**Selection:**
- Allow to place multiple cursors by alt+drag mouse. Currently it's only possible by clicking, adding dragging to it would allow to selected a column faster.
- Allow to remove cursors by alt+clicking on already placed ones.

**Actions and shortcuts:**
- Add shortcut to duplicate selected lines/selected range with ctrl+d.
- Add function to quickly rename object and its references. Different from search-replace as it would detect the object instead of searching for plain matching text.

**Support:**
- Add support for config files and resources syntax highlighting and correct formatting checks.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Some features could be added in a plugin, some rather not, it would require research.

### Is there a reason why this should be core and not an add-on in the asset library?

Text editing is tightly integrated into developing a game and it's helpful to it be as efficient as possible.

**Comment by AThousandShips** [comment]
> Allow to place multiple cursors by alt+drag mouse. 

* Tracked in: https://github.com/godotengine/godot-proposals/discussions/7004

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.
PS: In the future, please open one proposal per requested feature. Do not put several features that can be implemented independently from each other in a single proposal.

> Add function to quickly rename object and its references. Different from search-replace as it would detect the object instead of searching for plain matching text.

- This is already being tracked in https://github.com/godotengine/godot-proposals/issues/899.

> Add support for config files and resources syntax highlighting and correct formatting checks.

This is being implemented by https://github.com/godotengine/godot/pull/77972. Note that ConfigFile and TSCN/TRES don't have a style guide though, as these are not expected to be modified by hand.

**Comment by kitbdev** [comment]
> Add shortcut to duplicate selected lines/selected range with ctrl+d.

Implemented in https://github.com/godotengine/godot/pull/66553

**Reply by aXu-AP** [reply]
To add: you can change the shortcuts in editor. I personally make these modifications whenever I use Godot on new machine:
* Duplicate lines -> `Ctrl D`
* Add selection next occurence -> `Ctrl Shift D`
* Duplicate selection -> `none`

**Reply by kitbdev** [reply]
I like to set duplicate lines to `Shift Alt Down` since that's the default in VSCode. That shortcut by default is set to Script Editor script list move down, but that's not very useful since changing the order doesn't affect anything and it gets sorted by default anyway.
